File: src/lua/graphics/Scene.lua
-- https://github.com/Archzenvald/Shido
-- MIT license (see LICENSE, src/c/core.h or src/lua/core.lua)

local ffi = require("ffi")
local API = require("shido.API")
local core = require("shido.core")

-- C API
API.defineHandle("Scene")
ffi.cdef[[
shido_SceneRef* shido_Scene_new();
]]
local L = ffi.load("shido")

local graphics = shido.graphics

-- Window

function graphics.newScene()
  local ref = L.shido_Scene_new()
  if ref == nil then core.error() end
  return API.claimFreeRef(ref)
end

-- methods
local Scene = {}

API.implementHandle("Scene", Scene)

File: src/c/graphics/Scene.cpp
// https://github.com/Archzenvald/Shido
// MIT license (see LICENSE, src/c/core.h or src/lua/core.lua)

#include "Scene.hpp"

SHIDO_HANDLE_CPP(Scene)

shido_Scene::~shido_Scene()
{
  shido::Graphics::get().engine->destroy(handle);
}

shido_SceneRef* shido_Scene_new()
{
  filament::Scene *handle = shido::Graphics::get().engine->createScene();
  if(!handle){
    shido::error("Filament Scene creation failed.");
    return nullptr;
  }
  return new shido_SceneRef{std::make_shared<shido_Scene>(handle)};
}
